### PR TITLE
Switch to iRODS HTTP API

### LINKED ISSUE #41: update to use the new iRODS HTTP API
new HTTP API: https://github.com/irods/irods_client_http_api

---

Two options...

A) convert all the existing calls from using the REST API to the HTTP API
 - depends on the irods_demo being updated as well

OR

B) add the ability for this library to talk to the HTTP API, in addition to the REST API
 - add configuration option to select/target which API
 - every function has to learn how to hit both

It's not clear whether there will be any demand for B) once most people upgrade to the HTTP API (and iRODS 4.3.1+).

### PR DESCRIPTION
**Pull request by MartinSchobben**
This updates rirods to version 0.2.0 which is based on iRODS HTTP API 0.2.0. 

Things to consider:

- ~Wait for new HTTP API release (0.2.0), this will e.g. remove count parameter from write endpoint~
- ~Documentation has only been roughly adapted for now~
- ~Only the old rirods 0.1.2 functionality has now been implemented with some minor addition, such as, the server-info endpoint~

### CODE REVIEW

File: R/collections.R [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
 #'
 #' @param logical_path Path to the data object or collection to remove.
 #' @param force Whether the data object or collection should be deleted
 #'    permanently. If `FALSE`, it is sent to the trash collection. Defaults to
 #'   `TRUE`.
-#' @param recursive If a collection is provided, whether its contents should also be
-#'    removed. If a collection is not empty and `recursive` is `FALSE`, it cannot
-#'    be deleted. Defaults to `FALSE`.
+#' @param recursive If a collection is provided, whether its contents should
+#'     also bec removed. If a collection is not empty and `recursive` is `FALSE`
```
Comment (montesmariana) [inline R/collections.R:11]: "also be" instead of "also bec"

**Comment by MartinSchobben** [comment]
I think jsonlite should be moved to `Imports` in DESCRIPTION. It somehow evaded the R CMD checks.

File: R/data-objects.R [hunk 2]
```diff
@@ -11,10 +11,11 @@
 #' @param count Maximum number of bytes to write. Defaults to 2000.
 #' @param truncate Whether to truncate the object when opening it. Defaults to
 #'  `TRUE`.
+#' @param append Whether to append the objects.
+#' @param ticket A valid iRODS ticket string. Defaults to `NULL`.
 #' @param verbose Whether to print information about the HTTP request and
 #'  response. Defaults to `FALSE`.
-#' @param overwrite Whether the file in iRODS should be overwritten if it
-#'  exists. Defaults to `FALSE`.
+#' @param overwrite Deprecated. Use `truncate`.
```
Comment (montesmariana) [inline R/data-objects.R:18]: How does `truncate` relate to `overwrite`?
And what happens when `append=FALSE`, an overwrite? So the user should make sure first that the data object exists if they want to avoid overwriting?

Comment (montesmariana) [reply R/data-objects.R:18]: (I see how `truncate` relates in the code below, but it's not clear from this documentation what `truncate` means...

Comment (MartinSchobben) [reply R/data-objects.R:18]: Since `count` will be removed from the end-point, `truncate` will have the same effect as `overwrite`. Or do I miss here something. Could also rename `truncate`  to `overwrite`  and deprecate `truncate`.

**Comment by MartinSchobben** [comment]
@montesmariana. Thanks for already looking into this! There are a lot of consideration again with the new API. Perhaps we can once go through this over a video call.

File: R/data-objects.R (hunk 2 above)
Comment (montesmariana) [reply R/data-objects.R:18]: Yes, i think from the user's perspective `overwrite` is more clear (even if under the hood it is sent as `truncate`)
And what happens then with `append=FALSE`?

**Comment by korydraughn** [comment]
Given the number of files being modified in this PR and me not being familiar with R, are there files we can/should ignore?

Are there files you want us to focus on?

**Comment by MartinSchobben** [comment]
> Are there files you want us to focus on?

The division of commits is as follows. 

- The first commit involves the actual changes in R (only 23 files). 
- The second commit involves documentation of the previous R files.
- The third commit are the unit tests.
- The fourth commit are the HTTP snapshots for the former unit tests. They capture the HTTP response, so that one can test even without internet or iRODS server.
- Lastly, automatically generated README

I mainly wanted to let you (iRODS) now that this is underway. I will in the next week discuss with Mariana the finer details of the R package update.

**Comment by korydraughn** [comment]
Got it. And thanks for the explanation. I somehow didn't think to look at the commits.

Good to hear things appear to work.

We just released v0.2.0 of the HTTP API. It's available via docker hub. See  https://hub.docker.com/r/irods/irods_http_api/tags.

**Comment by MartinSchobben** [comment]
This is then also directly available through the irods_demo. Right?

**Comment by MartinSchobben** [comment]
I answered that my self. But there seems to be a problem:

From running docker compose up:

```
irods-client-http-api-1   | Traceback (most recent call last):
irods-client-http-api-1   |   File "<string>", line 1, in <module>
irods-client-http-api-1   |   File "/usr/local/lib/python3.10/dist-packages/jsonschema/validators.py", line 1312, in validate
irods-client-http-api-1   |     raise error
irods-client-http-api-1   | jsonschema.exceptions.ValidationError: 'max_size_of_request_body_in_bytes' is a required property
irods-client-http-api-1   | 
irods-client-http-api-1   | Failed validating 'required' in schema['properties']['http_server']['properties']['requests']:
irods-client-http-api-1   |     {'properties': {'max_size_of_request_body_in_bytes': {'minimum': 0,
irods-client-http-api-1   |                                                           'type': 'integer'},
irods-client-http-api-1   |                     'threads': {'minimum': 1, 'type': 'integer'},
irods-client-http-api-1   |                     'timeout_in_seconds': {'minimum': 1,
irods-client-http-api-1   |                                            'type': 'integer'}},
irods-client-http-api-1   |      'required': ['threads',
irods-client-http-api-1   |                   'max_size_of_request_body_in_bytes',
irods-client-http-api-1   |                   'timeout_in_seconds'],
irods-client-http-api-1   |      'type': 'object'}
irods-client-http-api-1   | 
irods-client-http-api-1   | On instance['http_server']['requests']:
irods-client-http-api-1   |     {'max_rbuffer_size_in_bytes': 8388608,
irods-client-http-api-1   |      'threads': 3,
irods-client-http-api-1   |      'timeout_in_seconds': 30}
irods-client-http-api-1   | Configuration failed validation.
irods-client-http-api-1   | Validating configuration file ...
irods-client-http-api-1   | No JSON schema file provided. Using default.
irods-client-http-api-1 exited with code 1
```

**Comment by MartinSchobben** [comment]
I guess this needs to be change now: https://github.com/irods/irods_demo/blob/main/irods_client_http_api/config.json

**Comment by korydraughn** [comment]
That's correct. The demo environment needs an update.

we'll get a fix in today.

**Comment by korydraughn** [comment]
Here's the PR that bumps the HTTP API to 0.2.0 in case you want to give it a try.

https://github.com/irods/irods_demo/pull/59

Let us know if it works for you.

File: DESCRIPTION [hunk 3]
```diff
@@ -51,5 +49,5 @@ Config/testthat/edition: 3
 Encoding: UTF-8
 Language: en-US
 Roxygen: list(markdown = TRUE)
-RoxygenNote: 7.2.3
-SystemRequirements: iRODS C++ REST API 0.9.3 (https://github.com/irods/irods_client_rest_cpp)
+RoxygenNote: 7.3.1
+SystemRequirements: iRODS C++ HTTP API 0.1.0 (<https://github.com/irods/irods_client_http_api>)
```
Comment (trel) [inline DESCRIPTION:53]: 0.2.0... once it's passing...

File: R/admin.R [hunk 4]
```diff
@@ -2,15 +2,10 @@
...
-#' @param target The subject of the action: user, zone, resource, childtoresc,
-#   childfromresc, token, group, rebalance, unusedAVUs, specificQuery.
-#' @param arg2 arg2
-#' @param arg3 arg3
-#' @param arg4 arg4
-#' @param arg5 arg5
-#' @param arg6 arg6
-#' @param arg7 arg7
+#' @param name Name of user to be added.
+#' @param password Password of to be added.
```
Comment (trel) [inline R/admin.R:6]: remove 'of'

File: R/admin.R [hunk 5]
```diff
@@ -2,15 +2,10 @@
...
+#' @param action The action: create_user, remove_user, or set_password.
+#' @param role Role of user: rodsuser or rodsadmin.
```
Comment (trel) [inline R/admin.R:8]: could also be ... 'groupadmin'?

File: R/admin.R [hunk 6]
```diff
@@ -30,40 +25,53 @@
...
-    arg2 = character(1),
-    arg3 = character(1),
-    arg4 = character(1),
-    arg5 = character(1),
-    arg6 = character(1),
-    arg7 = character(1),
+    name,
+    password = character(1),
+    action = c("create_user", "set_password", "remove_user"),
+    role = c("rodsuser", "rodsadmin"),
```
Comment (trel) [inline R/admin.R:40]: groupadmin too?

File: R/admin.R [hunk 7]
```diff
@@ -30,40 +25,53 @@
 #' iauth("rods", "rods")
```
Comment (trel) [inline R/admin.R:25]: line 22 above still mentions rest, rather than http

File: R/collections.R [hunk 8]
```diff
@@ -73,6 +83,8 @@ irm <- function(logical_path, force = TRUE, recursive = FALSE,
 #'   working directory (see [ipwd()]).
 #' @param create_parent_collections Whether parent collections should be created
 #'   when necessary. Defaults to `FALSE`.
+#' @param overwrite Whether the existin collection should be overwritten
```
Comment (trel) [inline R/collections.R:86]: existin -> existing

File: R/create-irods.R [hunk 9]
```diff
@@ -9,15 +9,17 @@
 #' it follows platform conventions (see also [rappdirs::user_config_dir()]).
 #'
 #' @param host URL of host.
-#' @param zone_path Path to the zone of the iRODS server
-#'    (e.g., "/tempZone/home").
+#' @param zone_path Depracated
```
Comment (trel) [inline R/create-irods.R:12]: Depracated -> Deprecated

File: R/data-objects.R (hunk 2 above)
Comment (trel) [reply R/data-objects.R:18]: @korydraughn please chime in here.... i think overwrite makes it a bit clearer, as well.

File: R/data-objects.R [hunk 10]
```diff
@@ -119,80 +147,203 @@ isaveRDS <- function(
...
+      append = append,
+      verbose = verbose,
+      stream_index = NULL
+    )
+  }
+}
 
+# calculate chunk sizes
+calc_chunk_size <- function(x, count, max_number_of_parallel_write_streams) {
+  # stop if object size is less than  count
```
Comment (trel) [inline R/data-objects.R:278]: extra space between 'than' and 'count'.

File: R/irods-demo.R [hunk 11]
```diff
@@ -207,11 +206,12 @@ dry_run_irods <- function(user, pass, recreate) {
...
-  "irods_demo_irods-catalog-provider",
-  "irods_demo_irods-client-icommands",
-  "irods_demo_irods-client-rest-cpp",
-  "irods_demo_nginx-reverse-proxy"
+  "irods_demo-irods-catalog",
+  "irods_demo-irods-catalog-provider",
+  "irods_demo-irods-client-icommands",
+  "irods_demo-irods-client-rest-cpp",
+  "irods_demo-nginx-reverse-proxy",
+  "irods/irods_http_api"
```
Comment (trel) [inline R/irods-demo.R:214]: this doesn't seem correct...

File: R/irods-http.R [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+handle_irods_errors <- function(req) {
+  httr2::req_retry(
+    req,
+    max_tries = 3,
+    is_transient = ~ httr2::resp_status(.x) %in% c(429, 503)
+  ) |>
+    httr2::req_error(body = irods_errors) # show if nothing works
+}
+
+# iRODS rest api errors
```
Comment (trel) [inline R/irods-http.R:66]: rest -> http

File: R/irods-path.R [hunk 13]
```diff
@@ -1,81 +1,113 @@
...
-                               safely = TRUE) {
+make_irods_base_path <- function() {
+  check_irods_conf()
+  if (!is.null(find_irods_file("irods_zone"))) {
+    if (.rirods$user_role == "rodsuser") {
+      paste0("/", find_irods_file("irods_zone"), "/home/", .rirods$user)
+    } else if (.rirods$user_role == "rodsadmin") {
+      paste0("/", find_irods_file("irods_zone"), "/home")
+    } else {
+      stop("User role unkown", call. = FALSE)
```
Comment (trel) [inline R/irods-path.R:9]: unkown -> unknown

File: R/irods-path.R [hunk 14]
```diff
@@ -1,81 +1,113 @@
-get_absolute_lpath <- function(lpath, data_type = "data_object", open = "write",
-                               safely = TRUE) {
+make_irods_base_path <- function() {
+  check_irods_conf()
+  if (!is.null(find_irods_file("irods_zone"))) {
+    if (.rirods$user_role == "rodsuser") {
+      paste0("/", find_irods_file("irods_zone"), "/home/", .rirods$user)
+    } else if (.rirods$user_role == "rodsadmin") {
+      paste0("/", find_irods_file("irods_zone"), "/home")
+    } else {
```
Comment (trel) [inline R/irods-path.R:8]: could also be groupadmin

File: R/irods-path.R [hunk 15]
```diff
@@ -1,81 +1,113 @@
...
+      stop("User role unkown", call. = FALSE)
+    }
+
+  } else {
+    stop("iRODS zone unkown", call. = FALSE)
```
Comment (trel) [inline R/irods-path.R:13]: unkown -> unknown

File: R/metadata.R [hunk 16]
```diff
@@ -181,12 +190,14 @@ imeta <- function(
 #' Use SQL-like expressions to query data objects and collections based on different properties.
 #'
 #' @param query GeneralQuery for searching the iCAT database.
-#' @param limit Maximum number of rows to return. Defaults to 100.
-#' @param offset Number of rows to skip for paging. Defaults to 0.
+#' @param limit Maximum number of rows to return. Depracated.
```
Comment (trel) [inline R/metadata.R:193]: Depracated -> Deprecated

File: R/metadata.R [hunk 17]
```diff
@@ -181,12 +190,14 @@ imeta <- function(
...
+#' @param offset Number of rows to skip for paging. Depracated.
```
Comment (trel) [inline R/metadata.R:194]: Depracated -> Deprecated

File: R/metadata.R [hunk 18]
```diff
@@ -181,12 +190,14 @@ imeta <- function(
...
 #' @param type Type of query: 'general' (the default) or 'specific'.
 #' @param case_sensitive Whether the string matching in the query is case sensitive.
 #'   Defaults to `TRUE`.
 #' @param distinct Whether only distinct rows should be listed. Defaults to `TRUE`.
+#' @param parser Which parser to use: genquery1 or genquery2. Defaults to genquery1.
+#' @param sql_only Whether to use SQL syntax. Defaults to `FALSE`. Needs Genquery2.
```
Comment (trel) [inline R/metadata.R:200]: This parameter is the equivalent of 'dry-run'... it requests that the server generate and return to the caller the SQL that *would* have been run.  It does NOT execute the SQL.

`sql_only: An integer which instructs the API plugin to return SQL without executing it.`

From GenQuery2 directly...
https://github.com/irods/irods_api_plugin_genquery2/?#api-plugin

I suggest...

```
Whether to return SQL without executing it. Defaults to `FALSE`. Needs genquery2.
```

File: dev/Makefile [hunk 19]
```diff
@@ -9,11 +9,11 @@ all: update remove clean
...
-	git submodule update --init --force --recursive
+	git submodule update --init --force --recursive --remote
 # delete unnecessary docker modules
 remove:
 	rm -rf $(DOCKER_MODULES)
 # remove lines docker-compose
 clean:
-	head -n 43 '$(PATH_MODULES)/docker-compose.yml' > temp.yml
+	head -n 63 '$(PATH_MODULES)/docker-compose.yml' > temp.yml
 	mv temp.yml '$(PATH_MODULES)/docker-compose.yml'
```
Comment (trel) [inline dev/Makefile:19]: why is this necessary?

if you only stand up the services you need... you don't need to manipulate/change the .yml file... and this code won't have to be revisited if the source .yml files continue to change (which they will).

the 43 changing to 63 feels brittle...

File: inst/shell_scripts/iauth.sh [hunk 20]
```diff
@@ -1,5 +1,14 @@
...
+TOKEN=$(curl -X POST -H "Authorization: Bearer ${SECRETS}" "$3/authenticate")
+
+curl http://localhost:9001/irods-http-api/0.1.0/data-objects \
+    -H "Authorization: Bearer ${SECRETS}" \
+    --data-urlencode 'op=parallel-write-init' \
+    --data-urlencode 'lpath=<string>' \
+    --data-urlencode 'stream-count=<integer>' \
+    --data-urlencode 'truncate=<integer>' \
+    --data-urlencode 'append=<integer>' \
+    --data-urlencode 'ticket=<string>'
```
Comment (trel) [inline inst/shell_scripts/iauth.sh:14]: i'm not sure what this is doing to do... there are no values for lpath, and others...

**Review by trel — COMMENTED** [review]
still a lot of duplication, but we can address that over time.     the URL being *everywhere* feels very brittle.

File: R/admin.R (hunk 5 above)
Comment (MartinSchobben) [reply R/admin.R:8]: yes, will change that. Although this function is not exposed to the user, as we opted for user facing functionality only.

File: dev/Makefile (hunk 19 above)
Comment (MartinSchobben) [reply dev/Makefile:19]: If this is included in the CRAN submission, it will trigger multiple R CMD check errors and warning. For example, path length cannot be more 100 characters (`storing paths of more than 100 bytes is not portable: `). This has btw. not changed since earlier submission, where we also used a trimmed down version.

File: inst/shell_scripts/iauth.sh (hunk 20 above)
Comment (MartinSchobben) [reply inst/shell_scripts/iauth.sh:14]: I need to remove some files here. They where used for initial development and testing (comparing curl with libcurl + R output).

Comment (MartinSchobben) [reply inst/shell_scripts/iauth.sh:14]: Only some scripts are actually used in the package in relation with `use_irods_demo()`.

File: R/data-objects.R (hunk 2 above)
Comment (korydraughn) [reply R/data-objects.R:18]: `overwrite` should replace `truncate` and `append`.

`truncate` and `append` should not be exposed for a PUT _(assuming that is truly what's being modeled)_.

If you want to expose `truncate` and `append`, I recommend renaming the `iput` function to something that expresses the ability to write data at arbitrary locations within a data object. This also makes it obvious that the call will not trigger PUT specific policy.

> Yes, i think from the user's perspective `overwrite` is more clear (even if under the hood it is sent as `truncate`)
And what happens then with `append=FALSE`?

`truncate` and `append` are independent of each other. Truncation only applies when the data object is first opened. Appending only applies during **write** operations. When append is not enabled, write operations will write data at the location of the file position. When append is enabled, write operations will seek to the end of the file before writing data.

Comment (MartinSchobben) [reply R/data-objects.R:18]: Okay, I'll remove both `truncate` and `append` then an only use it internally

**Comment by MartinSchobben** [comment]
> still a lot of duplication, but we can address that over time. the URL being _everywhere_ feels very brittle.

Not sure what is meant. I presume this can only be documentation and testing, using the localhost and the associated URL. Pretty sure there is no hardcoded URL.

File: R/admin.R (hunk 6 above)
Comment (MartinSchobben) [reply R/admin.R:40]: Will add that as well.

File: R/metadata.R (hunk 18 above)
Comment (MartinSchobben) [reply R/metadata.R:200]: Did not understand the parameter. Will change this accordingly.

File: inst/shell_scripts/iauth.sh (hunk 20 above)
Comment (MartinSchobben) [reply inst/shell_scripts/iauth.sh]: I have removed those scripts now

**Comment by trel** [comment]
> > still a lot of duplication, but we can address that over time. the URL being _everywhere_ feels very brittle.
> 
> Not sure what is meant. I presume this can only be documentation and testing, using the localhost and the associated URL. Pretty sure there is no hardcoded URL.

I mean that I see the URL (and the version number) hard-coded in many places.   When we bump to 0.3.0 or 1.0.0 we'll need to touch everything again.   Was thinking perhaps there could be a single source of truth, and a variable that gets used everywhere.  Not critical, an optimization, perhaps.

**Comment by korydraughn** [comment]
@MartinSchobben We've merged https://github.com/irods/irods_demo/pull/59. The HTTP API is pinned to 0.2.0 in irods/irods_demo until the next release.

**Comment by MartinSchobben** [comment]
Well, it all seems to work with HTTP API 0.2.0. Also included now a dynamic way of generating the documentation in regards to the HTTP API version changes. Although the website does not like that nice for the Reference to functions: https://martinschobben.github.io/irods_client_library_rirods/index.html. Moreover I can't run one vignette against the irods_demo (local vs iRODS) due a problem with a dependency when capturing HTTP responses (so this does not relate to functionality of rirods).

**Comment by trel** [comment]
I see three references to 'REST' in the rendered page at https://martinschobben.github.io/irods_client_library_rirods/index.html

File: R/data-objects.R (hunk 2 above)
Comment (MartinSchobben) [reply R/data-objects.R]: I have removed the parameters `offset`, `truncate` and `append` from `iput`, `iget`, `isaveRDS` and `ireadRDS`. It will, however, still be allowed as input, but it triggers now a deprecation error. I think this fine level of control was anyway not needed for the targeted users, but maybe I am wrong with that. I considered that offset would similarly not be required when modelling `iput`.

File: R/irods-demo.R (hunk 11 above)
Comment (trel) [reply R/irods-demo.R:245]: oh... because irods_demo is not building its own image (but rather using the one already published), this is the container name for the running container?

i had expected to see `irods_demo-irods-client-http-api`

File: dev/Makefile (hunk 19 above)
Comment (trel) [reply dev/Makefile:19]: okay, just caught my eye.

**Comment by trel** [comment]
looks pretty good - consider squashing to 1-3 commits?

**Comment by MartinSchobben** [comment]
> looks pretty good - consider squashing to 1-3 commits?

Yes, I will squash back to original 4 commits. But I will first discuss the changes in more detail with Mariana. Also, I have been offered a review by users that might want to run/use R + iRODS in production. So maybe more changes are to follow.

**Comment by trel** [comment]
Of course - excellent.

File: R/irods-demo.R [hunk 21]
```diff
@@ -264,8 +283,10 @@ local_create_irods <- function(
...
 
   # switch to new iRODS project
-  create_irods(host, zone_path, overwrite = TRUE)
+  create_irods(host, overwrite = TRUE)
   withr::defer(unlink(path_to_irods_conf()), envir = env)
 
   invisible(dir)
 }
+
+rirods_host <- function() "http://localhost:9001/irods-http-api/0.2.0"
```
Comment (montesmariana) [inline R/irods-demo.R:292]: Why is this a function instead of just a variable?

Comment (MartinSchobben) [reply R/irods-demo.R]: this variable is now loaded on attach

**Comment by MartinSchobben** [comment]
The PR would now be ready. @montesmariana I included what we discussed. Except I have not used the `lifecycle` package for this submission yet. I intend to do this for subsequent submissions. One import question remains, what would be the appropriate version number bump for this release?

**Comment by trel** [comment]
> appropriate version number bump for this release

doesn't feel 1.0.0 yet... perhaps 0.2.0?

**Comment by MartinSchobben** [comment]
yes, I guess 0.2.0 is right

File: R/irods-conf.R [hunk 22]
```diff
@@ -17,20 +23,19 @@ create_config_dir <- function(path_to_config_dir) {
 path_to_irods_conf <- function() {
   path_to_config_dir <- rappdirs::user_config_dir("rirods")
   create_config_dir(path_to_config_dir)
```
Comment (montesmariana) [inline R/irods-conf.R:25]: In `dir.create(path_to_config_dir)` (in the definition of `create_config_dir()`) better to add `recursive=TRUE`. When testing this in our test environments some directory at the top didn't exist and therefore this function failed.

**Review by trel — COMMENTED** [review]
ah, you found the irods_demo -> irods-demo change...

File: R/data-objects.R [hunk 23]
```diff
@@ -89,7 +89,7 @@ iput <- function(
   reqs <-  local_to_irods(
     object = local_path,
     logical_path = logical_path,
-    count =  find_irods_file("max_size_of_request_body_in_bytes"),
+    count =  find_irods_file("max_size_of_request_body_in_bytes") * 0.99, # need to make smalller to make it work
```
[the remark this answers is not in the document]
Comment (MartinSchobben) [reply R/data-objects.R:92]: I just found this now. I don't understand why. I had not covered the case in the tests

[the remark this answers is not in the document]
Comment (trel) [reply R/data-objects.R:92]: okay, we'll eyeball it here as well.

[the remark this answers is not in the document]
Comment (MartinSchobben) [reply R/data-objects.R:92]: This would be the error: `Caused by error in `curl::curl_fetch_memory()`: ! Send failure: Broken pipe`

[the remark this answers is not in the document]
Comment (MartinSchobben) [reply R/data-objects.R]: This seemed to have been caused by my own confusion. I reverted the change back the http api configuration again dictates the maximum size of the request body without modification.

[the remark this answers is not in the document]
Comment (trel) [reply R/data-objects.R]: excellent.  resolving.

File: R/irods-conf.R (hunk 22 above)
Comment (MartinSchobben) [reply R/irods-conf.R:25]: Then the problem seems to be here: `rappdirs::user_config_dir("rirods")`, as this should ensure to find the config directory on most OS. But it can be mitigated with `recursive`.

Comment (MartinSchobben) [reply R/irods-conf.R:25]: I add the test below. Does this replicate the problem?

```
test_that("create iRODS configure directory works", {
  # store conf file
  tmp <- tempfile()
  file.copy(path_to_irods_conf(), tmp)
  unlink(path_to_irods_conf())
  pt <- rappdirs::user_config_dir("rirods")
  # add accidental none existing directory
  pt <- file.path(dirname(pt), "notexist", "rirods")
  create_config_dir(pt)
  unlink(pt)
  # restore conf file
  file.copy(tmp, path_to_irods_conf())
})
```

**Comment by MartinSchobben** [comment]
I think that the PR is nearing completion. See also the blog post draft: https://martinschobben.github.io/iRODS4R/posts/rirods-0-2-0/ for when it is accepted at CRAN. Please propose edits at the source:  https://github.com/MartinSchobben/iRODS4R/tree/dev-blog.

**Comment by trel** [comment]
> I think that the PR is nearing completion. See also the blog post draft: https://martinschobben.github.io/iRODS4R/posts/rirods-0-2-0/ for when it is accepted at CRAN. Please propose edits at the source: https://github.com/MartinSchobben/iRODS4R/tree/dev-blog.

please start a PR for the post, and then edits/comments can be added.

**Comment by MartinSchobben** [comment]
PR made: https://github.com/FAIReLABS/iRODS4R/pulls

File: R/irods-conf.R (hunk 22 above)
Comment (montesmariana) [reply R/irods-conf.R:25]: Yes, although I find it weird to write a test to test `create_config_dir()` in which you call a function that depends on `create_config_dir()` (nl. `path_to_irods_conf()`). Maybe something like this:

```r
test_that("create iRODS configuration directory works", {
  tmp_dir <- tempdir()
  expect_true(dir.exists(tmp_dir))
  unlink(tmp_dir, force=TRUE, recursive=TRUE)
  expect_false(dir.exists(tmp_dir)) # just making sure

  # test create_config_dir()
  dir_to_create <- file.path(tmp_dir, "rirods")
  create_config_dir(dir_to_create)
  expect_true(dir.exists(dir_to_create))

  # test path_to_irods_conf()
  expect_true(file.exists(path_to_irods_conf()))
})
```

Comment (MartinSchobben) [reply R/irods-conf.R:25]: The true problem here is `rappdirs::user_config_dir("rirods")`, where it seems you stumbled on an edge case not covered by this function. I think the following is claimed to be true for the OS shown below. Bu apparently this is not the case on your system.

```
# On unix
dir.exists(rappdirs::user_config_dir("something")) # FALSE
dir.exists(dirname(rappdirs::user_config_dir("something"))) # TRUE
# On windows
dir.exists(rappdirs::user_config_dir("something")) # FALSE
dir.exists(dirname(rappdirs::user_config_dir("something"))) # FALSE
dir.exists(dirname(dirname(rappdirs::user_config_dir("something")))) # TRUE
``` 
So I guess the best course of action is maybe to remove the whole function and just use `dir.create('path_to_conf', recursive=TRUE)'. This way the directory is always created independent of how many parents `rappdirs` claims there to be.

Comment (montesmariana) [reply R/irods-conf.R:25]: Yes, that makes sense.

Comment (MartinSchobben) [reply R/irods-conf.R:25]: I still reconsidered. I think this function is still nice to keep:

```
create_config_dir <- function(path_to_config_dir) {
  if (!dir.exists(path_to_config_dir)) {
    dir.create(path_to_config_dir, recursive = TRUE)
  }
}

```

The following test would test it (although maybe a bit overkill) without destroying the test-configuration file:

```
test_that("create iRODS configuration directory works", {
  tmp_dir <- rappdirs::user_config_dir("mock_config_dir")
  if (!dir.exists(tmp_dir))
    dir.create(tmp_dir, recursive = TRUE)
  expect_true(dir.exists(tmp_dir))
  unlink(tmp_dir, force=TRUE, recursive=TRUE)
  expect_false(dir.exists(tmp_dir)) # just making sure

  # test create_config_dir()
  expect_invisible(create_config_dir(tmp_dir))
  expect_true(dir.exists(tmp_dir))
})

```

**Comment by MartinSchobben** [comment]
> I asked @jspijker if they want to have a look at this PR at the RIVM, so maybe more changes will still follow.

Okay, the tests by RIVM will have to wait a bit. So, I guess we can proceed. Or will more testing proceed at KU Leuven @montesmariana?

**Comment by trel** [comment]
I see 9 commits.

Please confirm this is ready for merging as is.

**Comment by MartinSchobben** [comment]
This is ready to merge. Thanks

**Comment by trel** [comment]
We now have a candidate SHA for 0.2.0... 1301faffb6edcff11313aae2a78d2a55fc42b719

Please confirm that SHA is correct for CRAN and GitHub release.

**Comment by MartinSchobben** [comment]
I need to make another tiny PR. CRAN checks complained about a possible unsafe URL.

**Comment by trel** [comment]
got it.

**Comment by MartinSchobben** [comment]
It arrived at CRAN. https://cloud.r-project.org/web/packages/rirods/index.html

**Comment by trel** [comment]
Oh, excellent.   Which SHA shall we tag as 0.2.0?

**Comment by trel** [comment]
@MartinSchobben I assume defa2cab65e8943756676a063b75c5cb31dc298a

please confirm.

**Comment by MartinSchobben** [comment]
Yes, that's correct

**Comment by trel** [comment]
Got it - will tag and release here at GitHub.
